drv/sidecar-mainboard-controller/src/fan_modules.rs

```diff
@@ -0,0 +1,154 @@
+        self.presence
+    }
+
+    /// Get the presence status for a particular fan module
+    pub fn is_present(&self, idx: u8) -> bool {
+        self.presence[idx as usize]
+    }
+
+    /// Enable the HSC for a fan module
+    ///
+    /// The FPGA will automatically disable the HSC if a fan module is removed.
+    /// The module will need to be re-enabled once it is returned.
+    pub fn set_enable(&self, idx: u8) -> Result<(), FpgaError> {
+        self.fpga.write(
+            WriteOp::BitSet,
+            Addr::FAN0_STATE as u16 + idx as u16,
+            Reg::FAN0_STATE::ENABLE,
+        )?;
+        Ok(())
+    }
+
+    /// Disable the HSC for a fan module
+    ///
+    /// The FPGA will automatically disable the HSC if a fan module is removed.
+    /// The module will need to be re-enabled once it is returned.
+    pub fn clear_enable(&self, idx: u8) -> Result<(), FpgaError> {
+        self.fpga.write(
+            WriteOp::BitClear,
+            Addr::FAN0_STATE as u16 + idx as u16,
+            Reg::FAN0_STATE::ENABLE,
+        )?;
+        Ok(())
+    }
+
+    /// Change the state of a fan module's LED
+    ///
+    /// This function will only modify the state of an LED if the requested
+    /// module is present, otherwise it will keep it off.
+    pub fn set_led_state(&mut self, idx: u8, state: FanModuleLedState) {
+        self.led_state[idx as usize] = if self.is_present(idx) {
+            state
+        } else {
+            FanModuleLedState::Off
+        };
+    }
+
+    /// Turn LEDs on or off depending on their state
+    ///
+    /// The `blink_on` parameter is external state as to if a blinking LED
+    /// should be turned on or off, allowing for synchronization with other
+    /// LEDs which may be blinking.
+    pub fn update_leds(&self, blink_on: bool) -> Result<(), FpgaError> {
+        for (module, state) in self.led_state.iter().enumerate() {
+            match state {
+                FanModuleLedState::Off => self.led_off(module as u8)?,
+                FanModuleLedState::On => self.led_on(module as u8)?,
+                FanModuleLedState::Blink => {
+                    if blink_on {
+                        self.led_on(module as u8)?
+                    } else {
+                        self.led_off(module as u8)?
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    // private function to turn an LED on
+    fn led_on(&self, idx: u8) -> Result<(), FpgaError> {
+        self.fpga.write(
+            WriteOp::BitSet,
+            Addr::FAN0_STATE as u16 + idx as u16,
+            Reg::FAN0_STATE::LED,
+        )?;
+        Ok(())
+    }
+
+    // private function to turn an LED off
+    fn led_off(&self, idx: u8) -> Result<(), FpgaError> {
+        self.fpga.write(
+            WriteOp::BitClear,
+            Addr::FAN0_STATE as u16 + idx as u16,
+            Reg::FAN0_STATE::LED,
+        )?;
+        Ok(())
+    }
```

Comment:
Nit: I'd recommend a single `led_set` function to avoid duplicating the call to `fpga.write`; this also simplifies the case statement here.
```suggestion
                FanModuleLedState::Blink => {
                    self.led_set(module as u8, blink_on)?
                }
            }
        }
        Ok(())
    }

    // private function to turn an LED on
    fn led_on(&self, idx: u8) -> Result<(), FpgaError> {
        self.led_set(idx, true)
    }

    // private function to turn an LED off
    fn led_off(&self, idx: u8) -> Result<(), FpgaError> {
        self.led_set(idx, false)
    }

    // private function to turn an LED off
    fn led_set(&self, idx: u8, on: bool) -> Result<(), FpgaError> {
        self.fpga.write(
            if on {
                WriteOp::BitSet
            } else {
                WriteOp::BitClear
            },
            Addr::FAN0_STATE as u16 + idx as u16,
            Reg::FAN0_STATE::LED,
        )?;
        Ok(())
    }
```